mlir/include/mlir/Dialect/EmitC/IR/EmitC.td

```diff
@@ -288,6 +288,7 @@ def EmitC_CastOp : EmitC_Op<"cast",
   let arguments = (ins EmitCType:$source);
   let results = (outs EmitCType:$dest);
   let assemblyFormat = "$source attr-dict `:` type($source) `to` type($dest)";
+  let hasVerifier = 1;
```

Comment:
Do you want to remove `SameOperandsAndResultShape` in another PR?

Reply:
Yes, I do that subsequently in the PR that does `collapse/expand_shape`, where we allow cast between arrays at the same time.